### PR TITLE
#spotlight

### PR DESCRIPTION
**Pull request by joshxg**
Bit of a prototype to see how you feel. Found plenty of problems and things to do better which is good.

The maintainer is trying to make this component basically as flexible as possible and putting it a good amount of work. The code base is nice and small enough to be very approachable.

image

Package issues:
- Bug when there is no results 🤔  https://kbar.vercel.app/ this is happening here, but just reports an error in the console. In dev, the oh noes! screen pops up.
- Types need some help but aren't too bad. For example the package that's being used for keyword matching can accept a string array for keywords, but the type is only a string.
- I have no idea how the shortcuts work 🤔 
- The actions are memoized, so if you're using an action where the perform is a closure, you'll get funny results. I fixed this by making the function a `useRef`, but it's not ideal. Ideally would be able to pass a `deps` array in to the `  useRegisterActions` hook.

Issues with what i've done:
- I think ideally the host wouldn't know about navigating to other actions. For example, I wanted the `CustomerContainer` to register all the `customer` based navigation actions. However, that's not loaded until you navigate to it. As I right this now, I realise that I could have put this in the `CustomerNavigation` component.. 🤔 
- Actions are registered all over the place. Pros and cons, I think. Pro: It's super easy to register some stuff where the actual code is that does the action. Cons: They're scattered all over the place!
- The UI of it is horrid. Would be nice to have icons, like a `+` next to Create actions and some sort of navigation action next to navigation actions. Also subtitles. These are both got issues for em. Also better styling.

In general I *love* ❤️  it and think it is awesome. But in saying that, I love macos spotlight and even bought Alfred to enhance it! So I am incredibly bias. Would also need a bit more structure to make sure it doesn't go crazy, I suspect

### CODE REVIEW

File: packages/host/src/Host.tsx [hunk 1]
```diff
@@ -42,6 +50,122 @@ const Heading: FC<{ locale: string }> = props => (
...
 
+const CustomKBarSearch = styled(KBarSearch)(({ theme }) => ({
+  width: 500,
+  height: 50,
+  fontSize: 20,
+  backgroundColor: theme.palette.primary.main,
+  borderRadius: '5px',
+  ':focus-visible': {
+    outline: 'none',
+  },
```
Comment (mark-prins) [inline packages/host/src/Host.tsx:61]: I found the grey text hard to read - what about this?
```suggestion
  color: '#fff',
  width: 500,
  height: 50,
  fontSize: 20,
  backgroundColor: theme.palette.primary.main,
  borderRadius: '5px',
  ':focus-visible': {
    outline: 'none',
  },
  '::placeholder': {
    color: '#fff',
    opacity: 0.75,
  },
```

File: packages/host/src/Host.tsx [hunk 2]
```diff
@@ -51,74 +175,80 @@ const Host: FC = () => {
...
-                            <Route
-                              path={RouteBuilder.create(AppRoute.Messages)
-                                .addWildCard()
-                                .build()}
-                              element={
-                                <Heading locale={locale}>messages</Heading>
-                              }
-                            />
-                            <Route path="/" element={<DashboardService />} />
+              <CommandK>
```
Comment (mark-prins) [inline packages/host/src/Host.tsx:178]: more nesting than a bird sanctuary

**Review by mark-prins — APPROVED** [review]
🎉  Squeeeee!!! that is awesome

can see that being super helpful. I had a play around with the shortcuts, as the ones defined need a little refinement. I was playing with combinations too and that's also working well.

File: packages/host/src/Host.tsx (hunk 2 above)
Comment (joshxg) [reply packages/host/src/Host.tsx:183]: 🤣   Heh, yeah.

I ended up going with a `DependencyProvider` in cold chain to handle it all..

**Comment by mark-prins** [comment]
`go to ci 7` 🤯  we're not far away from a nice looking command line web app

**Pull request merged**